research/cifar10/evaluate_on_test.py

```diff
@@ -0,0 +1,748 @@
+                        f"Client Number {client_number}, Run folder: {run_folder_dir}: "
+                        f"Aggregated Server Best Model Test Performance: {agg_server_run_metric}",
+                    )
+                    best_server_agg_test_metrics.append(agg_server_run_metric)
+                    all_best_server_agg_test_metrics[run_folder_dir] += agg_server_run_metric / NUM_CLIENTS
+
+            if eval_last_global_model:
+                server_model = load_last_global_model(run_folder_dir)
+                server_run_metric = evaluate_cifar10_model(server_model, test_loader, metrics, device, is_apfl)
+                log(
+                    INFO,
+                    f"Client Number {client_number}, Run folder: {run_folder_dir}: "
+                    f"Server Last Model Test Performance: {server_run_metric}",
+                )
+                last_server_test_metrics.append(server_run_metric)
+                all_last_server_test_metrics[run_folder_dir] += (
+                    server_run_metric * num_examples["eval_set"] / aggregated_num_examples
+                )
+
+                if eval_over_aggregated_test_data:
+
+                    agg_server_run_metric = evaluate_cifar10_model(
+                        server_model, aggregated_test_loader, metrics, device, is_apfl
+                    )
+                    log(
+                        INFO,
+                        f"Client Number {client_number}, Run folder: {run_folder_dir}: "
+                        f"Aggregated Server Last Model Test Performance: {agg_server_run_metric}",
+                    )
+                    last_server_agg_test_metrics.append(agg_server_run_metric)
+                    all_last_server_agg_test_metrics[run_folder_dir] += agg_server_run_metric / NUM_CLIENTS
+
+        # Write the results for each client
+        if eval_best_pre_aggregation_local_models:
+            avg_test_metric, std_test_metric = get_metric_avg_std(pre_best_local_test_metrics)
+            log(
+                INFO,
+                f"""Client {client_number} Pre-aggregation Best Model Average Test
+                    Performance on own Data: {avg_test_metric}""",
+            )
+            log(
+                INFO,
+                f"""Client {client_number} Pre-aggregation Best Model St. Dev. Test
+                  Performance on own Data: {std_test_metric}""",
+            )
+            test_results[f"client_{client_number}_pre_best_model_local_avg"] = avg_test_metric
+            test_results[f"client_{client_number}_pre_best_model_local_std"] = std_test_metric
+
+            if eval_over_aggregated_test_data:
+                avg_test_metric, std_test_metric = get_metric_avg_std(pre_best_local_agg_test_metrics)
+                log(
+                    INFO,
+                    f"""Aggregated Client {client_number} Pre-aggregation Best Model Average Test
+                        Performance on own Data: {avg_test_metric}""",
```

Comment:
I might be wrong, but this isn't actually on the clients "own" data, but rather on the entirety of the test dataset? (similar for the logging below)

Reply:
Oh yeah wording is weird indeed. I can change it.